### PR TITLE
added versioned cloud config service

### PR DESCRIPTION
**Pull request by xh3b4sd**

### CODE REVIEW

File: service/cloudconfig/cloud_config.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+// service by best effort.
+func DefaultConfig() Config {
+	return Config{
+		// Dependencies.
+		Logger: nil,
+	}
+}
+
+// CloudConfig implements the cloud config service interface.
+type CloudConfig struct {
```
Comment (xh3b4sd) [inline service/cloudconfig/cloud_config.go]: This is the cloud config service to get the versioned templates. I copied that from the KVM operator to have a kind of equal interface. This just makes it work here.

File: service/cloudconfig/templates.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
+package cloudconfig
+
+const (
+	decryptTLSAssetsScriptTemplate = `#!/bin/bash -e
```
Comment (xh3b4sd) [inline service/cloudconfig/templates.go]: The templates just moved to here.

File: service/cloudconfig/v_0_1_0_master_template.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+func (e *v_0_1_0MasterExtension) Files() ([]k8scloudconfig.FileAsset, error) {
+	filesMeta := []k8scloudconfig.FileMetadata{
+		{
+			AssetContent: decryptTLSAssetsScriptTemplate,
+			Path:         "/opt/bin/decrypt-tls-assets",
+			Owner:        FileOwner,
+			Permissions:  FilePermission,
+		},
```
Comment (xh3b4sd) [inline service/cloudconfig/v_0_1_0_master_template.go]: The files and units and all that jazz just moved to here.

File: service/cloudconfig/v_0_1_0_worker_template.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
+package cloudconfig
+
+import (
+	"github.com/giantswarm/awstpr"
+	"github.com/giantswarm/certificatetpr"
+	k8scloudconfig "github.com/giantswarm/k8scloudconfig/v_0_1_0"
+	"github.com/giantswarm/microerror"
+)
+
+func v_0_1_0WorkerTemplate(customObject awstpr.CustomObject, certs certificatetpr.CompactTLSAssets) (string, error) {
```
Comment (xh3b4sd) [inline service/cloudconfig/v_0_1_0_worker_template.go]: This is the same as above but for the workers.

File: service/create/asg_stack.go [hunk 5]
```diff
@@ -179,7 +169,7 @@ func (s *Service) createASGStack(input asgStackInput) (bool, error) {
 	cloudconfigS3 := &awsresources.BucketObject{
 		AWSEntity: awsresources.AWSEntity{Clients: input.clients},
 		Name:      s.bucketObjectName(input.asgType),
-		Data:      cloudConfig,
+		Data:      template,
```
Comment (xh3b4sd) [inline service/create/asg_stack.go]: Numer 1 template magic.

File: service/create/cloud_config.go [hunk 6]
```diff
@@ -1,413 +0,0 @@
-package create
```
Comment (xh3b4sd) [inline service/create/cloud_config.go]: All this got replaced by the common cloud config service.

File: service/create/launch_configuration.go [hunk 7]
```diff
@@ -74,7 +68,7 @@ func (s *Service) createLaunchConfiguration(input launchConfigurationInput) (boo
 
 	cloudconfigS3 := &awsresources.BucketObject{
 		Name:      s.bucketObjectName(input.prefix),
-		Data:      cloudConfig,
+		Data:      template,
```
Comment (xh3b4sd) [inline service/create/launch_configuration.go]: Numer 2 template magic.

File: service/create/service.go [hunk 8]
```diff
@@ -357,7 +364,7 @@ func (s *Service) runMachine(input runMachineInput) (bool, string, error) {
 	var cloudconfigS3 resources.Resource
 	cloudconfigS3 = &awsresources.BucketObject{
 		Name:      s.bucketObjectName(input.prefix),
-		Data:      cloudConfig,
+		Data:      template,
```
Comment (xh3b4sd) [inline service/create/service.go]: Numer 3 template magic.

**Comment by xh3b4sd** [comment]
@puja108 you might want to test this.

**Comment by puja108** [comment]
So should we test this on gauss? I guess the current k8scc is already vendored in?

**Comment by xh3b4sd** [comment]
Please do so.

**Comment by xh3b4sd** [comment]
> LoL, why there are templates generated in 3 places?

I do not know. I just want to make the versioning work. There will be a heavy rewrite anyway when we go with cloudformations.

**Comment by puja108** [comment]
So the current issue we have is that route53 DNS entries are not created. The error message is a bit cryptic:

```{"caller":"github.com/giantswarm/aws-operator/service/create/service.go:493",“error”:“error processing cluster '3q31r': '[{/go/src/github.com/giantswarm/aws-operator/service/create/service.go:1051: could not create record set '[{/go/src/github.com/giantswarm/aws-operator/service/create/domains.go:52: error registering DNS record 'api.3q31r.k8s.gauss.eu-central-1.aws.gigantic.io'} {/go/src/github.com/giantswarm/aws-operator/resources/aws/record_set.go:35: } {/go/src/github.com/giantswarm/aws-operator/resources/aws/record_set.go:57: } {MalformedInput: ChangeBatch is not valid, expected ChangeResourceRecordSetsRequest\n\tstatus code: 400, request id: 16176987-a43e-11e7-a5d2-09d8d35a3b4f}]'} {execution failed}]'","time":"17-09-28 11:13:42.425"}```

Not sure what exactly is not valid, but I could imagine some templating stuff not having been propagated into the DNS creation part of aws-operator and validation of route53 batch entry failing because of that.

I can't go much deeper on this. @xh3b4sd mentioned yesterday: "What I would do it add logging around the code the error trace mentions and try to get a picture. And then also try with master to see a comparison."

@kopiczko you any other idea or insight into aws-operator?

**Comment by xh3b4sd** [comment]
> I could imagine some templating stuff not having been propagated into the DNS creation part of aws-operator and validation of route53 batch entry failing 

Do you think this is templating related? Can you point to the templates? A hint into the right direction would be super useful.

**Comment by puja108** [comment]
The failing part is this https://github.com/giantswarm/aws-operator/blob/master/service/create/domains.go#L40-L53

which is mainly connected to https://github.com/giantswarm/aws-operator/blob/master/resources/aws/record_set.go

**Comment by xh3b4sd** [comment]
This is not templating related.

**Comment by puja108** [comment]
Yeah, I just realized that, the template is coming in higher up: https://github.com/giantswarm/aws-operator/blob/master/service/create/domains.go#L14

File: glide.yaml [hunk 9]
```diff
@@ -1,19 +1,9 @@
...
 import:
 - package: github.com/aws/aws-sdk-go
-  version: master
-  subpackages:
-  - service/autoscaling
-  - service/s3
-  - service/elb
-  - service/route53
-  - awserr
+  version: v1.12.3
```
Comment (xh3b4sd) [inline glide.yaml]: There was a problem with the route53 package introduced in some master version. This caused the operator to fail when reconciling guest clusters. I pinned the latest version here. We should update at some point but for the external dependency some more control is a good thing.

**Pull request merged**